sec2_the-ruby-programming-language/2.7_arrays-iterators.rb

```diff
@@ -0,0 +1,91 @@
+# Arrays
+a = [1, 2, 3, 4, 5, 6, 7, 8, 9]
+p a       # [1, 2, 3, 4, 5, 6, 7, 8, 9]
+p a.last  # 9
+
+# Range
+x = 1..10 # inclusive
+p x
+p x.class # Range (class type)
+p x.to_a          # [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+p x.to_a.shuffle  # [2, 7, 8, 1, 10, 4, 3, 9, 5, 6]
+p x               # 1..10
+p x.to_a.shuffle! # [8, 5, 7, 4, 1, 6, 10, 3, 9, 2]
+p x               # 1..10
+
+z = x.to_a.shuffle!
```

Comment:
Why need `!`?

Reply:
Good question, now I refer back I don't think `!` is needed because when you shuffle `x` and assign to `z`, the `z` is already an array with shuffled elements, even when we do `x.to_a.shuffle!`, the `x` will still print out a range (`1..10`) instead of an array with shuffled elements.

Is there anything that I miss out here?